[DeviceMesh] Make the repr shorter when debug ENV not set

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #158822

Users want a shorter repr so this PR is trying to address that when TORCH_DISTRIBUTED_DEBUG is not set to DETAIL. Feedback and discussion is welcomed. Somehow I found that torch.set_printoptions is global, so I am hesitated to use it.

Now the print is like 

<img width="435" height="79" alt="image" src="https://github.com/user-attachments/assets/8f173287-7138-4fbe-a4a3-8483523b21e4" />

or 

<img width="485" height="104" alt="image" src="https://github.com/user-attachments/assets/21e34db9-56b5-47e2-9767-750d6105a273" />

or

<img width="675" height="97" alt="image" src="https://github.com/user-attachments/assets/53aa763e-7edd-4622-9cdb-37e2af8ec11f" />


cc @H-Huang @awgu @wanchaol @fegin @wz337 @wconstab @d4l3k @pragupta

fduwjj committed:
[DeviceMesh] Make the repr shorter when debug ENV not set

[ghstack-poisoned]

wz337 reviewed (approved):
Great improvement. Shorter repr definitely makes more sense as the world size becomes larger and larger.

wconstab reviewed (approved):
Lgtm. 
One possible improvement is to print it like 
DeviceMesh(DP=2, TP=4) (zip the names and shapes then format them somehow) in the case names are available. Maybe take a look at how namedtuple prints and use that for inspiration.

fduwjj committed:
Update on "[DeviceMesh] Make the repr shorter when debug ENV not set"


Users want a shorter repr so this PR is trying to address that when TORCH_DISTRIBUTED_DEBUG is not set to DETAIL. Feedback and discussion is welcomed. Somehow I found that torch.set_printoptions is global, so I am hesitated to use it.

Now the print is like 

<img width="435" height="79" alt="image" src="https://github.com/user-attachments/assets/8f173287-7138-4fbe-a4a3-8483523b21e4" />

or 

<img width="485" height="104" alt="image" src="https://github.com/user-attachments/assets/21e34db9-56b5-47e2-9767-750d6105a273" />

or

<img width="675" height="97" alt="image" src="https://github.com/user-attachments/assets/53aa763e-7edd-4622-9cdb-37e2af8ec11f" />


cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

xmfan reviewed (approved):
thanks!

kshitij12345 commented:
Before this PR, it was possible to rely on DeviceMesh `__repr__` to re-create the DeviceMesh but that is not the case post this PR.

As per Python Data-model, `__repr__` should if possible create a valid Python expression which can recreate the object - https://docs.python.org/3.15/reference/datamodel.html#object.__repr__

> Called by the [repr()](https://docs.python.org/3.15/library/functions.html#repr) built-in function to compute the “official” string representation of an object. If at all possible, this should look like a valid Python expression that could be used to recreate an object with the same value (given an appropriate environment). If this is not possible, a string of the form <...some useful description...> should be returned. The return value must be a string object. If a class defines [__repr__()](https://docs.python.org/3.15/reference/datamodel.html#object.__repr__) but not [__str__()](https://docs.python.org/3.15/reference/datamodel.html#object.__str__), then [__repr__()](https://docs.python.org/3.15/reference/datamodel.html#object.__repr__) is also used when an “informal” string representation of instances of that class is required.


I think for shorter description controlled by `TORCH_DISTRIBUTED_DEBUG`, we can instead add `__str__` method which is called by `print`.

https://docs.python.org/3.15/reference/datamodel.html#object.__str__

> Called by [str(object)](https://docs.python.org/3.15/library/stdtypes.html#str), the default [__format__()](https://docs.python.org/3.15/reference/datamodel.html#object.__format__) implementation, and the built-in function [print()](https://docs.python.org/3.15/library/functions.html#print), to compute the “informal” or nicely printable string representation of an object.
> The return value must be a [str](https://docs.python.org/3.15/library/stdtypes.html#textseq) object.
> 
> This method differs from [object.__repr__()](https://docs.python.org/3.15/reference/datamodel.html#object.__repr__) in that there is no expectation that [__str__()](https://docs.python.org/3.15/reference/datamodel.html#object.__str__) return a valid Python expression: a more convenient or concise representation can be used.
> 
> The default implementation defined by the built-in type [object](https://docs.python.org/3.15/library/functions.html#object) calls [object.__repr__()](https://docs.python.org/3.15/reference/datamodel.html#object.__repr__).

Wdyt?

fduwjj commented:
@kshitij12345 I see.. Let me think about it and get back to you on this.
